### PR TITLE
Add missing keys to translation

### PR DESCRIPTION
**Pull request by nvh95**
## Motivation
- When doing the translation (for example: Vietnamese), we have to "clone" the content of `en.json` file to `vi.json`. However, I don't know an efficient way to do it except manually copy each line from `en.json` to `vi.json`. This new script `add-missing-keys-to-other-language.js` will ease this process by automatically adding missing keys from `en.json` to `vi.json` (and leaving existing keys as they are), so contributors can translate directly from English to their language of choice. See the following image for more intuition:
Ease translation work with new script

### What does it do?

Example use: 
```bash
node scripts/front/add-missing-keys-to-other-language.js vi
```
- This script will accept an argument as a language (e.g: `vi` - Vietnamese). Then adding all missing keys from the corresponding `en.json` in the same folder to `vi.json`. Contributors can translate from English in `vi.json` directly, they don't have to manually copy and paste from `en.json` to `vi.json`.
- One bonus point is that it also reorder the translation files

### Why is it needed?

- To help contributors translate easily
- To encourage contributors to translate Strapi to their language, which can potentially help Strapi expands the market globally.

### Others
- I used this script to help me translate some items from English to Vietnamese and I love it, it boosts my productivity and happiness. The translation is included in this PR

### Demo
- https://imgur.com/a/O558Zr9

### CODE REVIEW

**Comment by nvh95** [comment]
Demo gif:
https://imgur.com/a/O558Zr9

File: scripts/front/add-missing-keys-to-other-languages.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      if (currentTranslationFileJSON[current]) {
+        acc[current] = currentTranslationFileJSON[current];
+      } else {
+        acc[current] = mainTranslationFileJSON[current];
+      }
+
+      return acc;
+    }, {});
+    await fs.writeJson(filePath, updatedFile, { spaces: 2 });
+    console.log('Added missing keys to ', filePath);
```
Comment (vincentbpro) [inline scripts/front/add-missing-keys-to-other-languages.js:33]: I suggest removing space to add consistency
```suggestion
    console.log('Added missing keys to', filePath);
```

File: scripts/front/add-missing-keys-to-other-languages.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }, []);
+  console.log('List of files to add missing keys', translationFiles, '\n');
+
+  // For each file run addMissingKeyForSingleFile
+  translationFiles.forEach(addMissingKeyForSingleFile);
+};
+
+if (process.argv.length < 3) {
+  console.log(
+    'Please provide a language.For examples:\nnode scripts/front/add-missing-keys-to-other-languages.js vi'
```
Comment (vincentbpro) [inline scripts/front/add-missing-keys-to-other-languages.js:62]: Same here for consistency
```suggestion
    'Please provide a language. For example:\nnode scripts/front/add-missing-keys-to-other-languages.js vi'
```

**Review by vincentbpro — CHANGES_REQUESTED** [review]
Hello @nvh95 thanks for your contribution!
It's a cool addition, I have left a few comments :)

**Comment by nvh95** [comment]
@vincentbpro I pushed the changes. Please take a look. Thanks.

**Review by vincentbpro — DISMISSED** [review]
LGTM :)

**Comment by vincentbpro** [comment]
@soupette do you want to double check or are we good to go?

File: scripts/front/add-missing-keys-to-other-languages.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  const corePackageDirs = await glob('packages/core/*');
+  const pluginsPackageDirs = await glob('packages/plugins/*');
+  const packageDirs = [...corePackageDirs, ...pluginsPackageDirs];
+  const pathToTranslationsFolder = ['admin', 'src', 'translations'];
+
+  const translationFiles = packageDirs
+    .filter(dir => {
+      return fs.existsSync(join(dir, ...pathToTranslationsFolder, `${lang}.json`));
+    })
+    .reduce((acc, dir) => {
```
Comment (soupette) [inline scripts/front/add-missing-keys-to-other-languages.js:51]: Here you can use a `.map` instead.

```js
.map(dit => join(dir, ...pathTo, `${lang}.json`)
```

**Review by soupette — COMMENTED** [review]
Thank you @nvh95 this is a great addition!

I just have a small comment but it's fine if you leave the work as it is.

However would you mind adding some tests before we merge it?

**Comment by nvh95** [comment]
> Thank you @nvh95 this is a great addition!
> 
> I just have a small comment but it's fine if you leave the work as it is.
> 
> However would you mind adding some tests before we merge it?

@soupette Sure. I will tag you after I update some tests

File: scripts/front/__tests__/add-missing-keys-to-other-language.test.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+const fs = require('fs-extra');
```
Comment (soupette) [inline scripts/front/__tests__/add-missing-keys-to-other-language.test.js:3]: I think we should mock this so we are not actually writing files.

**Review by soupette — CHANGES_REQUESTED** [review]

File: scripts/front/__tests__/add-missing-keys-to-other-language.test.js (hunk 4 above)
Comment (nvh95) [reply scripts/front/__tests__/add-missing-keys-to-other-language.test.js:3]: @soupette I intentionally want to write files to disk, after the test finishes, I revert it as before the test executes. I always try to make the test as real as possible. 
For the problem of infinite loop (jest watch), I found a solution by configuring [watchPathIgnorePatterns](https://jestjs.io/docs/configuration#watchpathignorepatterns-arraystring) to ignore some test data files (`en.json`, `vi.json`). I think it's safe to ignore all json files, I don't think there will be a test that actually wants to watch json file. If there is, we can always update `watchPathIgnorePatterns`
```diff
// jest.base-config.js
+    watchPathIgnorePatterns: ['<rootDir>/.*\\.json'],
```


`yarn test:unit --watch scripts/front/__tests__/add-missing-keys-to-other-language.test.js`  before update the config (https://github.com/strapi/strapi/pull/12812/commits/d85a387b24e40e63c86cf03ae4ad486704de57e1)
Mar-15-2022 15-25-37
`yarn test:unit --watch scripts/front/__tests__/add-missing-keys-to-other-language.test.js`  after update the config:
Mar-15-2022 15-23-23

Comment (soupette) [reply scripts/front/__tests__/add-missing-keys-to-other-language.test.js:3]: Sorry but I disagree tests should not write files to disk want we want to test is your algorithm not that the fs lib is working correctly.

Comment (nvh95) [reply scripts/front/__tests__/add-missing-keys-to-other-language.test.js:3]: @soupette What I implemented here is a kind of e2e test. Just want to ignore all the implementation details and test if it works in a real-world example.
However, I completely understand that maybe by the convention of Strapi team, we expect a unit test on the algorithm i this scenario. Or we want tests to run faster by eliminating the File I/O operations here. I can look into that and update the test not to write files to disk later and mention you.
Thanks for your feedback

Comment (nvh95) [reply scripts/front/__tests__/add-missing-keys-to-other-language.test.js:3]: @soupette I just updated to extract the algorithm to a separated function `updateMissingKeysToJSON` and test it. There is no file writing in that function. Can you please take a look? Thanks.

**Review by soupette — APPROVED** [review]

**Pull request merged**